### ISSUE TITLE
Dataset understanding

### ISSUE THREAD

**Issue by Sujie1528** [yilundu/nerflow#8]
In the pouring dataset `/nerf_synthetic/test`, there's a mix of r_x.png and r_x_flow.png (where x is the index), should the r_x_flow.png be interpreted as the optical flow map? How are they generated? Where can I find a script to generate them?
In the Gibson dataset, `/gibson_dataset/gibson_test_100_smaller_range_0_6_0_0_0_4`, there are 5 kinds of data: d_x_flow.npy, d_x.npy, flow_x.npy, r_x.png, r_x_flow.png. Again how to interpret them and how to compute them based on the raw data (which I assume are RGB and depth images)? Where can I find a script to generate them?
Also what is the data in the `render_linear` folder?

TL; DR: What is the structure of each dataset? How to interpret the files? Are there scripts to generate them?
Thanks in advance!

**Comment by yilundu** [comment]
Hi,

Thanks for the question -- please see the comment below about the descriptions of each portion of the dataset -- feel to ask more questions if its not clear:

For the pouring dataset r_x.png and r_x_flow.png is a RGB image of the scene at a particular viewpoint at timestep t and at a subsequent timestep (t + delta) These images are generated using blender -- I can share the code if its helpful.

For the Gibson dataset r_x.png and r_x_flow.png are the same as the pouring setting. d_x.npy and d_x_flow.npy are the depths of the images in the same setting. flow_x.npy is the optical flow between the r_x.png and r_x_flow.png as computed by raft -- I can share the code also if its helpful.

**Comment by Sujie1528** [comment]
Please share the script that generates these files. Thank you so much!

**Comment by Sujie1528** [comment]
Also, in the dataset, how are the transforms.json files generated? Can you also share the script? Thank you.

**Comment by megatran** [comment]
Could you also clarify on how  you got the `dimensions` and `offset` parameters in `transforms.json`? Would any type of depth image (e.g., not HDR) still work? Thank you!

**Comment by yilundu** [comment]
Hi yes -- in the process of sharing the scripts -- its been a while so I need to find them. Any type of depth image should work. Offset represent the time offset different between the image captured at timestep A and timestep B, while dimensions corresponds to the rough size of motion. The precise value of dimensions shouldn't be that important -- you can likely use the same numbers currently in the gibson dataset for other scenes.

**Comment by megatran** [comment]
Thanks for clarifying!

Looking at the snippet of the Gibson dataset below (`frame r_0` and `r_0_flow`), I had some follow-up questions

1. The `timestep` is different `-0.20425746480289586` vs `-0.0042574648028958495`, does that mean the `r_0_flow` happens *before* the original `r_0`?

2. If `offset` denotes the time offset between `r_0` and `r_o_flow`, then it seems to not add up to the `timestep` between `r_0` and `r_0_flow`. Nor `r0 (-0.20425746480289586)` and `r1  (0.674204720217909)`. Or perhaps in simulation, the timestep has a different meaning than the discrete timestep like in the pouring dataset. And since the offset is a 3D vector, does this imply a time offset in three x,y, and z dimensions?

3. Is `dimensions` in some sort of uv plane (or normalized 0-1 space)? 

- _When you said it represents the size of motion, does it mean after running the flow estimation, you determined where the motion happens in the scene (for example in Gibson, the most colorful one would be the robot since it's the only thing that moves so the dimension here would be the bounding box surrounding the robot in the flow map?). In the pouring dataset, the dimensions field is fixed so I'm guessing the bowl isn't moving like the robot in Gibson_

4. Do you remember how many cameras were there in the Gibson data collection? 


Snippet from `transforms.json`

```
  "camera_angle_x": 90,
  "frames": [
    {
      "file_path": "gibson_train_300_smaller_range_0_6_0_0_0_4/r_0.png",
      "depth_path": "gibson_train_300_smaller_range_0_6_0_0_0_4/d_0.npy",
      "transform_matrix": [
        [
          -0.3131185472011566,
          0.9497140645980835,
          0,
          0
        ],
        [
          -0.7150620818138123,
          -0.2357543408870697,
          0.6581079959869385,
          -2.9802322387695312e-8
        ],
        [
          0.6250143647193909,
          0.2060658037662506,
          0.7529235482215881,
          -1
        ],
        [
          0,
          0,
          0,
          1
        ]
      ],
      "timestep": -0.20425746480289586,
      "offset": [
        [
          -0.5215060281828234,
          0.02986138027956564,
          0.21829999999993976
        ]
      ],
      "dimensions": [
        0.365855,
        0.358,
        0.47320000000006024
      ]
    },
    {
      "file_path": "gibson_train_300_smaller_range_0_6_0_0_0_4/r_0_flow.png",
      "depth_path": "gibson_train_300_smaller_range_0_6_0_0_0_4/d_0_flow.npy",
      "transform_matrix": [
        [
          -0.3131185472011566,
          0.9497140645980835,
          0,
          0
        ],
        [
          -0.7150620818138123,
          -0.2357543408870697,
          0.6581079959869385,
          -2.9802322387695312e-8
        ],
        [
          0.6250143647193909,
          0.2060658037662506,
          0.7529235482215881,
          -1
        ],
        [
          0,
          0,
          0,
          1
        ]
      ],
      "timestep": -0.0042574648028958495,
      "offset": [
        [
          -0.3265060281828235,
          0.059861380279565654,
          0.21829999999993976
        ]
      ],
      "dimensions": [
        0.365855,
        0.358,
        0.47320000000006024
      ]
    },
    {
      "file_path": "gibson_train_300_smaller_range_0_6_0_0_0_4/r_1.png",
      "depth_path": "gibson_train_300_smaller_range_0_6_0_0_0_4/d_1.npy",
      "transform_matrix": [
        [
          0.20447584986686707,
          0.9788715839385986,
          0,
          -1.4901161193847656e-8
        ],
        [
          -0.6380413770675659,
          0.13328003883361816,
          0.7583795785903931,
          -2.9802322387695312e-8
        ],
        [
          0.7423562407493591,
          -0.15507030487060547,
          0.6518131494522095,
          -1
        ],
        [
          0,
          0,
          0,
          1
        ]
      ],
      "timestep": 0.6742047202179098,
      "offset": [
        [
          0.33499460221246213,
          0.1616307080326865,
          0.21829999999993976
        ]
      ],
      "dimensions": [
        0.365855,
        0.358,
        0.47320000000006024
      ]
    },
    {
      "file_path": "gibson_train_300_smaller_range_0_6_0_0_0_4/r_1_flow.png",
      "depth_path": "gibson_train_300_smaller_range_0_6_0_0_0_4/d_1_flow.npy",
      "transform_matrix": [
        [
          0.20447584986686707,
          0.9788715839385986,
          0,
          -1.4901161193847656e-8
        ],
        [
          -0.6380413770675659,
          0.13328003883361816,
          0.7583795785903931,
          -2.9802322387695312e-8
        ],
        [
          0.7423562407493591,
          -0.15507030487060547,
          0.6518131494522095,
          -1
        ],
        [
          0,
          0,
          0,
          1
        ]
      ],
      "timestep": 0.8742047202179097,
      "offset": [
        [
          0.5299946022124621,
          0.19163070803268653,
          0.21829999999993976
        ]
      ],
      "dimensions": [
        0.36585499999999993,
        0.358,
        0.47320000000006024
      ]
    },
```

**Comment by yilundu** [comment]
Ahh sorry its been a long time since I've used this code. Please see the following answers:

1) Yes the timestep difference can be both before and after
2) Sorry my bad -- that is incorrect, actually offset denotes the movement of the robot, I don't think its used in the code itself.
3) Dimensions is in x,y,z space and denote the size of the robot
4) I believe each camera is randomly sampled in the upper hemisphere (since this is a simulation) -- so in principle each frame is a different camera.

I'm trying to find the code I used to generate the data which should hopefully make everything much clearer.

**Comment by yilundu** [comment]
See below for the code to render one of these datasets:

```
import yaml
from gibson2.core.physics.robot_locomotors import Turtlebot, Humanoid
from gibson2.core.simulator import Simulator
from gibson2.core.physics.scene import BuildingScene, StadiumScene
from gibson2.utils.utils import parse_config
import pytest
import pybullet as p
import numpy as np
import os
import gibson2
from gibson2.utils.assets_utils import download_assets, download_demo_data
from skimage.io import imsave
import imageio
import json
from scipy.ndimage import gaussian_filter1d
from gibson2.external.pybullet_tools.utils import get_aabb
import random
import math


def listify_matrix(matrix):
    matrix_list = []
    for row in matrix:
        matrix_list.append(list([float(r) for r in row]))
    return matrix_list

if __name__ == "__main__":


    VIEWS = 150
       
#     prefix = "gibson_train_dual_cams_300_-0_5_0_5_-0_1_0_2"
    prefix = "gibson_train_stereo_cams_300_-0_5_0_5_-0_1_0_2_r_1_0_z_0_6_pi_1_6"
    
    RENDER_DUAL_CAM = True
    RENDER_FLOW = True

    if not os.path.isdir(prefix):
        os.makedirs(prefix)

    config = parse_config(os.path.join(gibson2.assets_path, 'example_configs/turtlebot_interactive_demo.yaml'))
    s = Simulator(mode='headless', image_width=800, image_height=800)
    scene = BuildingScene('Rs_interactive', is_interactive=True)
    # scene = BuildingScene('Rs')
    s.import_scene(scene)
    config['robot_scale'] = 1.0
    turtlebot = Turtlebot(config)
    s.import_robot(turtlebot)

    frames = {}
    ims = []

    offset = np.random.randn(350, 3) * 0.02
    offset = gaussian_filter1d(offset, 0.5, axis=0)

    pos = np.array([1, 0, 0.5])

    data = {}
    frames = []
    data['camera_angle_x'] = s.renderer.vertical_fov

    timesteps = np.linspace(-1, 1, VIEWS)

    objects = []

    start_pos = np.array([-0.5, -0.1, 0.0])
    end_pos = np.array([0.5, 0.2, 0.0])
#     start_pos = np.array([-0.8, 0.0, 0.0])
#     end_pos = np.array([0.5, 0.2, 0.0])

    scale = 1.0
    cam_range = 0.2
    angles = np.linspace(-math.pi/6, math.pi/6, VIEWS)
    for i in range(VIEWS):
        print(i)
        
        timestep = timesteps[i]
        
        rescale_timestep = (timestep + 1) / 2.

        target_pos = start_pos + (end_pos - start_pos) * rescale_timestep
        turtlebot.set_position(target_pos)
        s.sync()

#         x_1 = np.random.uniform(0.6-cam_range, 0.6+cam_range)
#         x_2 = np.random.uniform(0.0-cam_range, 0.0+cam_range)
#         x_3 = np.random.uniform(0.4-cam_range, 0.4+cam_range)
#         x = np.array([x_1, x_2, x_3])
#         #x = np.random.uniform(-1, 1, (3,))
#         #cam_sampled = x
#         x[2] = np.abs(x[2]) + 0.2
#         x_norm = x / np.linalg.norm(x)
#         x_norm = scale * x_norm
#         s.renderer.set_camera(x_norm, [0, 0, 0], [0, 0, 1])
#         x = np.random.uniform(-1, 1, (3,))
        x_3 = 0.6
#         theta = (2*math.pi)/VIEWS * i
        theta = angles[i]
        
        x_1 = math.sqrt(1.0 - x_3**2) * math.cos(theta)
        x_2 = math.sqrt(1.0 - x_3**2) * math.sin(theta)
        x = np.array([x_1, x_2, x_3])
#         x[2] = np.abs(x[2]) + 0.2
#         x_norm = x / np.linalg.norm(x)
        print("x_norm:", np.linalg.norm(x))

#         x_norm = scale * x_norm
#         s.renderer.set_camera(x_norm, [0, 0, 0], [0, 0, 1])
        s.renderer.set_camera(x, [0, 0, 0], [0, 0, 1])
        view_matrix = s.renderer.V.copy()

        im, depth = s.renderer.render(modes=('rgb', '3d'))
        low_bound, high_bound = get_aabb(s.renderer.instances[-1].robot.robot_body.bodies[0])

        path = "{}/r_{}.png".format(prefix, 2*i)
        #path = "{}/r_{}_cam_{}_{}_{}.png".format(prefix, i, cam_sampled[0], cam_sampled[1], cam_sampled[2])
        depth_path = "{}/d_{}.npy".format(prefix, 2*i)
        
        frame_data = {}
        frame_data['file_path'] = path
        frame_data['depth_path'] = depth_path
        frame_data['transform_matrix'] = listify_matrix(view_matrix)
        frame_data['timestep'] = timestep
        frame_data['offset'] =  [list(low_bound + high_bound / 2)]
        frame_data['dimensions'] = (list((high_bound - low_bound)))
        frames.append(frame_data)
        imsave(path, im)
        np.save(depth_path, depth)
        
        if RENDER_DUAL_CAM:
            theta_prime = theta + math.pi / 6.
        
            x_1_prime = math.sqrt(1.0 - x_3**2) * math.cos(theta_prime)
            x_2_prime = math.sqrt(1.0 - x_3**2) * math.sin(theta_prime)
            x_3 = 0.6
            x = np.array([x_1_prime, x_2_prime, x_3])
            s.renderer.set_camera(x, [0, 0, 0], [0, 0, 1])
            view_matrix = s.renderer.V.copy()

            im, depth = s.renderer.render(modes=('rgb', '3d'))
            low_bound, high_bound = get_aabb(s.renderer.instances[-1].robot.robot_body.bodies[0])

            path = "{}/r_{}.png".format(prefix, 2*i+1)
            #path = "{}/r_{}_cam_{}_{}_{}.png".format(prefix, i, cam_sampled[0], cam_sampled[1], cam_sampled[2])
            depth_path = "{}/d_{}.npy".format(prefix, 2*i+1)
            frame_data = {}
            frame_data['file_path'] = path
            frame_data['depth_path'] = depth_path
            frame_data['transform_matrix'] = listify_matrix(view_matrix)
            frame_data['timestep'] = timestep
            frame_data['offset'] =  [list(low_bound + high_bound / 2)]
            frame_data['dimensions'] = (list((high_bound - low_bound)))
            frames.append(frame_data)
            imsave(path, im)
            np.save(depth_path, depth)

        if RENDER_FLOW:
            x = np.array([x_1, x_2, x_3])
            s.renderer.set_camera(x, [0, 0, 0], [0, 0, 1])
            view_matrix = s.renderer.V.copy()

            if random.randint(0, 1) < 0.5:
                t_offset = 0.2
            else:
                t_offset = -0.2
            path = "{}/r_{}_flow.png".format(prefix, 2*i)
            #path = "{}/r_{}_flow_cam_{}_{}_{}.png".format(prefix, i, cam_sampled[0], cam_sampled[1], cam_sampled[2])
            depth_path = "{}/d_{}_flow.npy".format(prefix, 2*i)
            target_pos = start_pos + (end_pos - start_pos) * rescale_timestep

            target_pos = target_pos + t_offset * (end_pos - start_pos) / 2.
            turtlebot.set_position(target_pos)

            s.sync()
            low_bound, high_bound = get_aabb(s.renderer.instances[-1].robot.robot_body.bodies[0])
            im, depth = s.renderer.render(modes=('rgb', '3d'))

            imsave(path, im)
            np.save(depth_path, depth)

            frame_data = {}
            frame_data['file_path'] = path
            frame_data['depth_path'] = depth_path
            frame_data['transform_matrix'] = listify_matrix(view_matrix)
            frame_data['timestep'] = timestep + t_offset
            frame_data['offset'] =  [list(low_bound + high_bound / 2)]
            frame_data['dimensions'] = (list(high_bound - low_bound))
            frames.append(frame_data)
            
            if RENDER_DUAL_CAM:
                x = np.array([x_1_prime, x_2_prime, x_3])
                s.renderer.set_camera(x, [0, 0, 0], [0, 0, 1])
                view_matrix = s.renderer.V.copy()

                if random.randint(0, 1) < 0.5:
                    t_offset = 0.2
                else:
                    t_offset = -0.2
                path = "{}/r_{}_flow.png".format(prefix, 2*i+1)
                #path = "{}/r_{}_flow_cam_{}_{}_{}.png".format(prefix, i, cam_sampled[0], cam_sampled[1], cam_sampled[2])
                depth_path = "{}/d_{}_flow.npy".format(prefix, 2*i+1)
                target_pos = start_pos + (end_pos - start_pos) * rescale_timestep

                target_pos = target_pos + t_offset * (end_pos - start_pos) / 2.
                turtlebot.set_position(target_pos)

                s.sync()
                low_bound, high_bound = get_aabb(s.renderer.instances[-1].robot.robot_body.bodies[0])
                im, depth = s.renderer.render(modes=('rgb', '3d'))

                imsave(path, im)
                np.save(depth_path, depth)

                frame_data = {}
                frame_data['file_path'] = path
                frame_data['depth_path'] = depth_path
                frame_data['transform_matrix'] = listify_matrix(view_matrix)
                frame_data['timestep'] = timestep + t_offset
                frame_data['offset'] =  [list(low_bound + high_bound / 2)]
                frame_data['dimensions'] = (list(high_bound - low_bound))
                frames.append(frame_data)


    data['frames'] = frames
    json_object = json.dumps(data, indent=4)
    with open("{}/transform_train.json".format(prefix), "w") as f:
        f.write(json_object)
    # imsave("test.png", im)
```

**Comment by yilundu** [comment]
Here is the code used to render the pouring dataset (in Blender)

```
"""
https://blender.stackexchange.com/questions/28480/how-do-i-force-the-netrender-to-render-everything-on-gpu/28513#28513
blender -b file.blend -E CYCLES -t 0 -o //file -P script.py

time /viscam/u/mguo95/opt/blender-2.82a-linux64/blender -b bouncing_ball_v3.blend -E CYCLES -t 0 -o ani_frames/ -noaudio --python render_from_blend.py

"""
# import util
import os
import bpy
import json
import math
import random
import _cycles
import argparse
import numpy as np

# from ball_dataset_cfg import START_FRAME, END_FRAME
START_FRAME = 0
END_FRAME = 449
import json

def save_json(path, data):
    with open(path, 'w') as f:
        json.dump(data, f, indent=4)


def load_json(path):
    with open(path, "r") as f:
        metadata = json.load(f)
    return metadata


def gen_rand_rot():
    rot = np.random.uniform(0, 1, size=3) * (1,0,2*np.pi)
    rot[0] = np.abs(np.arccos(1 - 2 * rot[0]) - np.pi/2)
    return rot

def gen_rand_rot_t(t, ts):
    rot = np.random.uniform(0, 1, size=3) * (1,0,2*np.pi)
    rot[2] = 2 * np.pi * float(t) / float(ts)
    rot[0] = np.abs(np.arccos(1 - 2 * rot[0]) - np.pi/2)
    return rot

def gen_rand_rot_t_camera(t, ts, frame_id):
    rot = np.random.uniform(0, 1, size=3) * (1,0,2*np.pi)

    if frame_id % 2 == 0:
        rot[2] = 2 * np.pi * float(t) / float(ts)
    else:
        rot[2] = 2 * np.pi * float(t) / float(ts) + np.pi / 5

    # if random.randint(0, 1):
    rot[0] = TRAJ[t]
    # rot[0] = np.random.uniform(0, 2*np.pi)
    # else:
    #     rot[0] = TRAJ_2[t]

    return rot

rots = [gen_rand_rot() for i in range(3)]
TYPE = "camera"
# TYPE = "normal"
# TYPE = "correlated"
# TYPE = "camera"

FLOW_OFFSET_FRAC = 0.1  # 0.25; yilun's offset / (1 - (-1)) = offset / 2
# N_RANDOM_EXAMPLES = 1000
N_RANDOM_EXAMPLES = 100
N_TILES = 512
# RESOLUTION = 400
RESOLUTION = 400
RESOLUTION_PERCENT = 100
# RENDER_ROT = np.array([0.06637931, 0., 1.98244318])  # Ball dataset
# RENDER_ROT = np.array([0.236488, 0., 3.31160372])  # Bball dataset
RENDER_ROT = np.array([1.1, 0., 3.31160372])  # Bball dataset
TRAJ = np.ones(END_FRAME + 1) * 1.1
TRAJ_2 = np.ones(END_FRAME + 1) * 1.1

dt = np.pi / (END_FRAME - START_FRAME) * 32

# for i in range(START_FRAME + 1, END_FRAME):
#     perturb = (np.random.uniform(-1, 1) * dt)
#     TRAJ[i] = min(max(TRAJ[i-1] * 0.7 + perturb * 0.3, 0.05), np.pi/2)
# 
# for i in range(START_FRAME + 1, END_FRAME):
#     perturb = (np.random.uniform(-1, 1) * dt)
#     TRAJ_2[i] = min(max(TRAJ_2[i-1] * 0.7 + perturb * 0.3, 0.05), np.pi/2)

CAM_LOCATION = (0, 10, 0)  # (30.9, -36.9, 22.0)
CAM_CLIP_END = 30.0
DATASET2INFO = {
    "ball": {
        "object_name": "Sphere"
    },
    "bball": {
        "object_name": "NBA BASKETBALL"
    }
}


def main():
    # render_mode = argv[0]
    # dataset = argv[1]
    depth_node = setup_render_settings()
    render_frames('random', 'bball', depth_node)


def render_frames(render_mode, dataset, depth_node):
    """https://blender.stackexchange.com/questions/17839/python-render-specific-frames"""

    ts2data = generate_object_data(dataset)

    if render_mode == "random":
        examples = generate_random_examples(n=N_RANDOM_EXAMPLES)
    elif render_mode == "render":
        examples = generate_render_split()
    else:
        raise ValueError("Invalid render mode: {args.render_mode}")

    generate_dataset(ts2data, examples, depth_node)


def generate_object_data(dataset):
    scene = bpy.context.scene
    object_name = DATASET2INFO[dataset]["object_name"]
    # obj = scene.objects[object_name]

    ts2data = {}
    for ts in range(START_FRAME, END_FRAME + 1):
        # Set the animation frame.
        scene.frame_set(ts)

        # Store the object properties.
        ts2data[ts] = {
            'offset': [0.1, 0.1, 0.1],
            'dimensions': [0.1, 0.1, 0.1]
        }

    return ts2data


def generate_random_examples(n):
    timesteps = []
    # for _ in range(n):

    #     sampled_ts = random.randint(START_FRAME, END_FRAME)  # Inclusive
    #     # while True:
    #     #     sampled_ts = random.randint(START_FRAME, END_FRAME)  # Inclusive
    #     #     if sampled_ts % 30 == 2:
    #     #         break

    #     timesteps.append(sampled_ts)
    # sorted_sampled_timesteps = sorted(timesteps)
    sorted_sampled_timesteps = list(range(START_FRAME, END_FRAME+1))
    timesteps = sorted_sampled_timesteps
    timesteps = list(np.repeat(timesteps, 2))
    # print(sorted_sampled_timesteps)
    # timesteps = list(range(START_FRAME, END_FRAME, 10))

    # with open("/data/vision/billf/scratch/yilundu/nerf-pytorch/data/nerf_synthetic/pouring_dataset_cvpr/transforms_video.json", "r") as f:
    #     meta = json.load(f)

    # frames = meta['frames']
    # frames = frames[::4][::10]
    examples = []

    # for frame_id, frame in enumerate(frames):
    #     ts = frame['timestep']
    #     rot = frame['rot']
    #     examples.append((frame_id, ts, None, rot))

    for frame_id, ts in enumerate(timesteps):
        ts = int(ts)
        # This for correlated views on a circle

        # timestep_idx = ts % 3
        # rot = rots[timestep_idx]

        if TYPE == "correlated":
            rot = gen_rand_rot_t(ts, END_FRAME)
        elif TYPE == "camera":
            rot = gen_rand_rot_t_camera(ts, END_FRAME, frame_id)
        elif TYPE == "normal":
            rot = gen_rand_rot()

        print(rot)
    #     # This is for non correlated_view on a circle
        flow_ts = get_rand_flow_ts(ts)
        examples.append((frame_id, ts, flow_ts, rot))

    examples = examples[::2][::10]

    return examples


def get_rand_flow_ts(ts):
    n_frames = END_FRAME - START_FRAME + 1
    offset = n_frames * FLOW_OFFSET_FRAC
    if random.randint(0, 1) < 0.5:
        offset *= -1
    flow_ts = ts + int(offset)
    flow_ts = max(START_FRAME, min(flow_ts, END_FRAME))
    return flow_ts


def generate_render_split():
    examples = []
    for ts in range(START_FRAME, END_FRAME + 1, 10):
        frame_id = ts
        rot = RENDER_ROT
        examples.append((frame_id, ts, None, rot))
    return examples


def generate_dataset(ts2data, examples, depth_node):
    scene = bpy.context.scene
    out_dir = scene.render.filepath
    # assert not os.path.exists(out_dir)


    # Extract the scene objects.
    cam = scene.objects["Camera"]

    for i in range(20):
        bpy.data.worlds['World'].cycles.sample_as_light = True
        bpy.context.scene.cycles.blur_glossy = 2.0
        # bpy.context.scene.cycles.samples = 512
        # bpy.context.scene.cycles.transparent_min_bounces = 6
        # bpy.context.scene.cycles.transparent_max_bounces = 6
        # bpy.context.scene.cycles.samples = 16

        # Low resolution version
        bpy.context.scene.cycles.samples = 2
        bpy.context.scene.cycles.transparent_min_bounces = 1
        bpy.context.scene.cycles.transparent_max_bounces = 2
        bpy.context.scene.cycles.film_transparent = True

        bpy.ops.object.lamp_add(type='SUN')
        lamp2 = bpy.data.lamps['Sun']
        lamp2.use_specular = True
        lamp2.energy = 100.0
        bpy.ops.object.lamp_add(type='POINT', radius=30.0, location=(0, 0.0, 10.0))

    # Set up the camera to track an imaginary object at the origin.
    dummy = parent_obj_to_camera(cam)
    setup_cam(cam, dummy)

    idx, metadata = load_metadata(out_dir)

    if metadata is None:
        metadata = {
            "camera_angle_x": cam.data.angle_x,
            "frames": []
        }

    flow_metadata = {
        "camera_angle_x": cam.data.angle_x,
        "frames": []
    }


    for counter, (frame_id, ts, flow_ts, rot) in enumerate(examples):
        counter = counter + idx + 1
        # Set the frame to render.
        scene.frame_set(ts)

        # Set the rotation of the center object, which updates the camera pose.
        dummy.rotation_euler = rot

        # Render and save to path.
        path = os.path.join(out_dir, "r_{}".format(counter))
        scene.render.filepath = path
        depth_node.file_slots[0].path = "r_{}".format(counter)
        depth_path = os.path.join("depth", "r_{}".format(counter) + "{:04}".format(ts) + ".hdr")

        bpy.ops.render.render(write_still=True)

        # Store frame metadata.
        meta = construct_metadata(path, depth_path, cam, ts, ts2data, rot)
        metadata["frames"].append(meta)
        save_metadata(metadata, out_dir)

        # Store flow metadata
        if flow_ts is not None:
            scene.frame_set(flow_ts)

            path = os.path.join(out_dir, "r_{}_flow".format(counter))
            scene.render.filepath = path
            depth_node.file_slots[0].path = "r_{}_flow".format(counter)
            depth_path = os.path.join("depth", "r_{}_flow".format(counter) + "{:04}".format(ts) + ".hdr")
            flow_meta = construct_metadata(path, depth_path, cam, flow_ts, ts2data, rot)
            metadata["frames"].append(flow_meta)

            bpy.ops.render.render(write_still=True)
            save_metadata(metadata, out_dir)

        if counter > N_RANDOM_EXAMPLES:
            break


def construct_metadata(path, depth_path, cam, ts, ts2data, rot):
    frame_data = {
        'file_path': path + ".png",
        'depth_path': depth_path,
        'transform_matrix': listify_matrix(cam.matrix_world),
        'timestep': ts,
        'rot': list(rot),
        'offset': ts2data[ts]["offset"],
        'dimensions': ts2data[ts]["dimensions"]
    }
    return frame_data

def load_metadata(out_dir):
    path = os.path.join(out_dir, "transforms.json")

    if os.path.exists(path):
        metadata = load_json(path)
        idx = len(metadata['frames']) // 2
    else:
        idx = 0
        metadata = None

    return idx, metadata

def save_metadata(metadata, out_dir, flow=False):
    if flow:
        fname = "transforms_flow.json"
    else:
        fname = "transforms.json"
    path = os.path.join(out_dir, fname)
    save_json(path=path, data=metadata)


def parent_obj_to_camera(b_camera):
    b_empty = bpy.data.objects.new("Empty", None)
    b_empty.location = (0, 0, 0)
    b_camera.parent = b_empty  # setup parenting
    return b_empty


def setup_cam(cam, dummy):
    cam.location = CAM_LOCATION
    cam.data.clip_end = CAM_CLIP_END
    cam_constraint = cam.constraints.new(type='TRACK_TO')
    cam_constraint.track_axis = 'TRACK_NEGATIVE_Z'
    cam_constraint.up_axis = 'UP_Y'
    cam_constraint.target = dummy



def listify_matrix(matrix):
    matrix_list = []
    for row in matrix:
        matrix_list.append(list(row))
    return matrix_list


def setup_render_settings():
    avail_devices = _cycles.available_devices()
    print(avail_devices)

    prop = bpy.context.user_preferences.addons['cycles'].preferences

    # prop.get_devices(prop.compute_device_type)
    prop.compute_device_type = 'CUDA'

    for device in prop.devices:
        if device.type == 'CUDA':
            print('device: ', device)
            device.use = True

    bpy.context.user_preferences.addons['cycles'].preferences.compute_device_type = 'CUDA'
    bpy.context.user_preferences.addons['cycles'].preferences.devices[0].use= True

    scene = bpy.context.scene
    scene.use_nodes = True
    scene.render.layers['RenderLayer'].use_pass_object_index = True
    scene.cycles.device = 'GPU'

    scene.render.tile_x = N_TILES
    scene.render.tile_y = N_TILES

    scene.render.resolution_x = RESOLUTION
    scene.render.resolution_y = RESOLUTION
    scene.render.resolution_percentage = RESOLUTION_PERCENT

    scene.render.image_settings.file_format = "PNG"

    # Add generation of a depth image
    nodes = scene.node_tree.nodes
    links = scene.node_tree.links

    # map_value_node = bpy.context.scene.node_tree.nodes.new("CompositorNodeMapValue")
    # map_value_node.size[0] = 1 / 20
    # map_value_node.size[0] = 1 / 20
    # links.new(nodes["Render Layers"].outputs["Z"], map_value_node.inputs[0])

    file_output_node = bpy.context.scene.node_tree.nodes.new('CompositorNodeOutputFile')
    file_output_node.base_path = "depth/"
    # use uint 16 for higher accuracy
    file_output_node.format.file_format = "HDR"

    links.new(nodes["Render Layers"].outputs["Z"], file_output_node.inputs[0])

    return file_output_node


if __name__ == '__main__':

    # parser = argparse.ArgumentParser()
    # parser.add_argument("render_mode", type=str, help="The render mode to run.")
    # args = parser.parse_args()
    import sys
    # argv = sys.argv
    # argv = argv[argv.index("--") + 1:]  # get all args after "--"
    main()
```
